### PR TITLE
Add support for raising if you set a tsconfig entry of target/module with the right setting in the root

### LINKED ISSUE #43293: Provide warnings/errors when putting tsconfig compilerOptions in the root of the tsconfig.json
# Suggestion

## 🔍 Search Terms

TSConfig root fields errors

## ✅ Viability Checklist

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

```json
{
    "target": "es2017",
    "module": "commonjs",
    "lib": [
      "esnext",
      "es2015",
      "es2016",
      "es2017"
    ] 
}
```

Should give some sort of feedback that you really want:

```json
{
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",
    "lib": [
      "esnext",
      "es2015",
      "es2016",
      "es2017"
    ] 
  }
}
```

## 📃 Motivating Example

I can't think of a reason why you might want to have these keys in the root of your tsconfig, open to ideas about why you might.

## 💻 Use Cases

New folks who accidentally set up the tsconfig wrong

### PR DESCRIPTION
**Pull request by orta**
Fixes #43293

### CODE REVIEW

**Review by andrewbranch — COMMENTED** [review]
This seems like a nice addition, but why limit it to `target` and `module`? Surely we have an array of all valid or common option names somewhere, right?

**Comment by orta** [comment]
Just wanted to do the least to avoid this wrath by using what is probably always gonna be set there:

> Other tools are allowed to use the top-level keys here, and do, so there's some danger of compat problems if in the future we add a compiler setting with the same name as a tool's setting

**Comment by andrewbranch** [comment]
What if you detect all compilerOptions keys in the root, but only issue an error if there is no `compilerOptions` object present at all? That way, third party tools could add any root key they want, even ones that share names with `compilerOptions` keys, as long as the user _also_ added `"compilerOptions": {}`.

**Comment by orta** [comment]
This idea is great, I'll switch it to that

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -2852,6 +2852,7 @@ namespace ts {
         let typeAcquisition: TypeAcquisition | undefined, typingOptionstypeAcquisition: TypeAcquisition | undefined;
         let watchOptions: WatchOptions | undefined;
         let extendedConfigPath: string | undefined;
+        const rootCompilerOptions: PropertyName[] = [];
```
Comment (andrewbranch) [inline src/compiler/commandLineParser.ts:2855]:
```suggestion
        let rootCompilerOptions: PropertyName[] | undefined;
```

File: src/compiler/commandLineParser.ts [hunk 2]
```diff
@@ -2890,22 +2891,12 @@ namespace ts {
...
-                    }
-                }
-                if (key === "module") {
-                    const moduleOption = commandOptionsWithoutBuild.find(opt => opt.name === "module")!;
-                    const possibleValidEntries = arrayFrom((moduleOption.type as Map<string>).keys());
-                    if (contains(possibleValidEntries, value)) {
-                        errors.push(createDiagnosticForNodeInSourceFile(sourceFile, keyNode, Diagnostics._0_should_be_set_inside_the_compilerOptions_object_of_the_tsconfig_json, "module"));
-                    }
+                if (find(commandOptionsWithoutBuild, (opt) => opt.name === key)) {
+                    rootCompilerOptions.push(keyNode);
```
Comment (andrewbranch) [inline src/compiler/commandLineParser.ts:2899]:
```suggestion
                    rootCompilerOptions = append(rootCompilerOptions, keyNode);
```

File: src/compiler/commandLineParser.ts [hunk 3]
```diff
@@ -2926,6 +2917,11 @@ namespace ts {
             }
         }
 
+        // eslint-disable-next-line no-in-operator
+        if (rootCompilerOptions.length && json && !("compilerOptions" in json)) {
```
Comment (andrewbranch) [inline src/compiler/commandLineParser.ts:2921]:
```suggestion
        if (rootCompilerOptions && json && !("compilerOptions" in json)) {
```

**Review by andrewbranch — APPROVED** [review]
All suggestions are for lazy array allocation

File: src/compiler/commandLineParser.ts [hunk 4]
```diff
@@ -2913,6 +2917,11 @@ namespace ts {
             }
         }
 
+        // eslint-disable-next-line no-in-operator
+        if (rootCompilerOptions && json && !("compilerOptions" in json)) {
```
Comment (DanielRosenwasser) [inline src/compiler/commandLineParser.ts:2921]: You could write `json.compilerOptions === undefined` and avoid the linter suppression, right?

**Review by DanielRosenwasser — APPROVED** [review]
Would be nice to just avoid the linter suppression.

File: src/compiler/commandLineParser.ts (hunk 4 above)
Comment (DanielRosenwasser) [inline src/compiler/commandLineParser.ts:2921]: if you just want to do it now

```suggestion
        if (rootCompilerOptions && json && json.compilerOptions === undefined) {
```

Comment (andrewbranch) [reply src/compiler/commandLineParser.ts:2921]: There’s also a `hasProperty` util

**Comment by orta** [comment]
OK, I've had a few looks over this will merge it at end of day

**Pull request merged**